### PR TITLE
Refactor RTP retransmission buffer in a separate and testable RTC::RetransmissionBuffer class

### LINKED ISSUE #975: consuming plain producer results in worker crash
I’m pushing video to mediasoup(3.11.4) producer with gstremer. There is no problems with it if there is no consumers for the producer. But after about 80 seconds of consuming (video play perfectly whole time) mediasoup worker process dies with `failed assertion 'this->buffer.size() <= MaxSeq': StorageItemBuffer contains more than 65535 entries`
<details> 
  <summary>Core dump:</summary>
<pre><code>
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
        set = {__val = {0, 94619480744673, 28202962353000, 56, 94619497407008, 94619497407720, 140736458774320, 94619497345440, 140283140088888, 94619480744716, 94619497031937, 28202962353000, 56,
            0, 18446744073709551615, 18446744073709551615}}
        pid = <optimized out>
        tid = <optimized out>
        ret = <optimized out>
#1  0x00007f96356aa7f1 in __GI_abort () at abort.c:79
        save_stage = 1
        act = {__sigaction_handler = {sa_handler = 0x1, sa_sigaction = 0x1}, sa_mask = {__val = {0, 0, 511101108348, 395136991342, 140283121853312, 94619497341264, 512, 140283121851456,
              4294934528, 94619569736784, 32769, 32705, 140283118359213, 390842024046, 140283121854080, 140283121836704}}, sa_flags = 896436269, sa_restorer = 0x560e55d448e0}
        sigs = {__val = {32, 0 <repeats 15 times>}}
        __cnt = <optimized out>
        __set = <optimized out>
        __cnt = <optimized out>
        __set = <optimized out>
#2  0x0000560e50497752 in RTC::RtpStreamSend::StorageItemBuffer::Insert(unsigned short, RTC::RtpStreamSend::StorageItem*) ()
No symbol table info available.
#3  0x0000560e50497b1e in RTC::RtpStreamSend::StorePacket(RTC::RtpPacket*, std::shared_ptr<RTC::RtpPacket>&) ()
No symbol table info available.
#4  0x0000560e50497e1a in RTC::RtpStreamSend::ReceivePacket(RTC::RtpPacket*, std::shared_ptr<RTC::RtpPacket>&) ()
No symbol table info available.
#5  0x0000560e504a9157 in RTC::SimpleConsumer::SendRtpPacket(RTC::RtpPacket*, std::shared_ptr<RTC::RtpPacket>&) ()
No symbol table info available.
#6  0x0000560e50479b3b in RTC::Router::OnTransportProducerRtpPacketReceived(RTC::Transport*, RTC::Producer*, RTC::RtpPacket*) ()
No symbol table info available.
#7  0x0000560e50465f33 in RTC::Producer::ReceiveRtpPacket(RTC::RtpPacket*) ()
No symbol table info available.
#8  0x0000560e504c0a91 in RTC::Transport::ReceiveRtpPacket(RTC::RtpPacket*) ()
No symbol table info available.
#9  0x0000560e50451ffb in RTC::PlainTransport::OnRtpDataReceived(RTC::TransportTuple*, unsigned char const*, unsigned long) ()
No symbol table info available.
#10 0x0000560e5045316e in non-virtual thunk to RTC::PlainTransport::OnUdpSocketPacketReceived(RTC::UdpSocket*, unsigned char const*, unsigned long, sockaddr const*) ()
No symbol table info available.
#11 0x0000560e50820cd3 in uv.udp_recvmmsg ()
No symbol table info available.
#12 0x0000560e50821a43 in uv.udp_io ()
No symbol table info available.
#13 0x0000560e50825286 in uv.io_poll ()
No symbol table info available.
#14 0x0000560e50817a06 in uv_run ()
No symbol table info available.
#15 0x0000560e503acdb9 in DepLibUV::RunLoop() ()
No symbol table info available.
#16 0x0000560e503ba751 in Worker::Worker(Channel::ChannelSocket*, PayloadChannel::PayloadChannelSocket*) ()
No symbol table info available.
#17 0x0000560e503ab2e6 in mediasoup_worker_run ()
No symbol table info available.
#18 0x0000560e503a9e74 in main ()
No symbol table info available.
</code></pre>

### PR DESCRIPTION
**Pull request by ibc**
A complete refactor of the retransmission buffer which now is a separate `RetransmissionBuffer` class.

- Fixes #975

### CODE REVIEW

**Comment by ibc** [comment]
@nazar-pc said in https://github.com/versatica/mediasoup/issues/975#issuecomment-1463665484:

> `idx <= static_cast<uint16_t>(this->buffer.size() - 1)` ensures this in one of the branches, it'll simply start overriding old values once we reach buffer size of elements.
>
> The last `else` branch is where this issue probably happens. I guess there was incorrect expectation about sequence numbers there and some packets do get processed out of order enough for sequence number to overflow and become below `this->startSeq`. I think similar check for buffer size needs to be added here and start overriding values instead of pushing beyond buffer size.
>
> It will result in messed up contents of packets when decoder tries to interpret it, but we for sure must not crash here.

I've added some logs in `RtpStreamSend::StorageItemBuffer::Insert()`:

```ts
	void RtpStreamSend::StorageItemBuffer::Insert(uint16_t seq, StorageItem* storageItem)
	{
		MS_DUMP("--- [seq:%" PRIu16 "]", seq);

		if (this->buffer.empty())
		{
			MS_DUMP("--- 1 [seq:%" PRIu16 "] buffer empty", seq);

			this->startSeq = seq;
			this->buffer.push_back(storageItem);
		}
		// Packet sequence number is higher than startSeq.
		else if (RTC::SeqManager<uint16_t>::IsSeqHigherThan(seq, this->startSeq))
		{
			MS_DUMP(
				"--- 2 [seq:%" PRIu16 "] IsSeqHigherThan(%" PRIu16 "%" PRIu16 ")",
				seq, seq, this->startSeq);

			auto idx{ static_cast<uint16_t>(seq - this->startSeq) };

			// Packet arrived out of order, so we already have a slot allocated for it.
			if (idx <= static_cast<uint16_t>(this->buffer.size() - 1))
			{
				MS_DUMP("--- 2.a [seq:%" PRIu16 "] packet out of order", seq);

				MS_ASSERT(this->buffer.at(idx) == nullptr, "must insert into empty slot");

				MS_DUMP("--- 2.a [seq:%" PRIu16 "] buffer[idx] = storageItem", seq);

				this->buffer[idx] = storageItem;
			}
			else
			{
				MS_DUMP("--- 2.b [seq:%" PRIu16 "] packet in order", seq);

				// Calculate how many elements would it be necessary to add when pushing new item
				// to the back of the deque.
				auto addToBack = static_cast<uint16_t>(seq - (this->startSeq + this->buffer.size() - 1));

				MS_DUMP("--- 2.b [seq:%" PRIu16 "] addToBack:%" PRIu16,
					seq, addToBack);

				// Packets can arrive out of order, add blank slots.
				for (uint16_t i{ 1 }; i < addToBack; ++i)
				{
					MS_DUMP(
						"--- 2.b [seq:%" PRIu16 "] buffer.push_back(nullptr)",
						seq);

					this->buffer.push_back(nullptr);
				}

				MS_DUMP(
					"--- 2.b [seq:%" PRIu16 "] buffer.push_back(storageItem)",
					seq);

				this->buffer.push_back(storageItem);
			}
		}
		// Packet sequence number is the same or lower than startSeq.
		else
		{
			MS_DUMP(
				"--- 3 [seq:%" PRIu16 "] ! IsSeqHigherThan(%" PRIu16 ", %" PRIu16 ")",
				seq, seq, this->startSeq);

			// Calculate how many elements would it be necessary to add when pushing new item
			// to the front of the deque.
			auto addToFront = static_cast<uint16_t>(this->startSeq - seq);

			MS_DUMP("--- 3 [seq:%" PRIu16 "] addToFront:%" PRIu16,
				seq, addToFront);

			// Packets can arrive out of order, add blank slots.
			for (uint16_t i{ 1 }; i < addToFront; ++i)
			{
				MS_DUMP(
					"--- 3 [seq:%" PRIu16 "] buffer.push_front(nullptr)",
					seq);

				this->buffer.push_front(nullptr);
			}

			MS_DUMP(
				"--- 3 [seq:%" PRIu16 "] buffer.push_front(storageItem)",
				seq);

			this->buffer.push_front(storageItem);
			this->startSeq = seq;
		}

		MS_DUMP(
			"--- 4 [seq:%" PRIu16 "] buffer.size:%zu",
			seq, this->buffer.size());

		MS_ASSERT(
		  this->buffer.size() <= MaxSeq,
		  "StorageItemBuffer contains more than %" PRIu16 " entries",
		  MaxSeq);
	}
```

By looking at the logs (see below) we are allowing new items to be introduced in the buffer without taking care of not filling it. And honestly I don't see anything exotic in these logs. Those are just ordered RTP packets (from time to time there are lost packets but the old ones never arrive). As you can see in the logs, just those prefixed with `--- 2.b` are executed and those are the responsible of the crash.

[packets.log](https://github.com/versatica/mediasoup/files/10957804/packets.log)

So here I'll ask a basic question:

**Should we check current buffer size within `RtpStreamSend::StorageItemBuffer::Insert()` method and override an older packet if needed? Ot should such a check be done in the caller side in `RtpStreamSend::StorePacket()` method?**

**Comment by ibc** [comment]
Another (unrelated) question:

```ts
		// Check if RTP packet is too old to be stored.
		if (this->storageItemBuffer.GetBufferSize() > 0)
		{
			auto* storageItem = this->storageItemBuffer.GetFirst();

			// Processing RTP packet is older than first one.
			if (RTC::SeqManager<uint32_t>::IsSeqLowerThan(packet->GetTimestamp(), storageItem->timestamp))
			{
				const uint32_t diffTs{ storageItem->timestamp - packet->GetTimestamp() };

				// RTP packet is older than the retransmission buffer size.
				if (static_cast<uint32_t>(diffTs * 1000 / this->params.clockRate) >= this->retransmissionBufferSize)
				{
					return;
				}
			}
		}
```

Here we are storing a packet if it's not older than oldest stored packet + retransmission buffer size. Shouldn't we just store it if it's not older than **newest** stored packet + retransmission buffer size?

**Comment by jmillan** [comment]
> Shouldn't we just store it if it's not older than newest stored packet + retransmission buffer size?

correct :+1:

**Comment by ibc** [comment]
> > Shouldn't we just store it if it's not older than newest stored packet + retransmission buffer size?
> 
> correct 👍

I'm adding this `GetLast()` method. Is it ok? I think so but it's unclear to me whether it's guaranteed that the last element in the buffer list always holds a packet:

```c++
	RtpStreamSend::StorageItem* RtpStreamSend::StorageItemBuffer::GetFirst() const
	{
		auto* storageItem = this->Get(this->startSeq);

		MS_ASSERT(storageItem, "first storage item is missing");
		MS_ASSERT(storageItem->packet, "storage item does not contain original packet");

		return storageItem;
	}

	RtpStreamSend::StorageItem* RtpStreamSend::StorageItemBuffer::GetLast() const
	{
		auto* storageItem = this->Get(this->startSeq + this->buffer.size() - 1);

		MS_ASSERT(storageItem, "last storage item is missing");
		MS_ASSERT(storageItem->packet, "storage item does not contain original packet");

		return storageItem;
	}
```

**Comment by jmillan** [comment]
> Should we check current buffer size within RtpStreamSend::StorageItemBuffer::Insert() method and override an older packet if needed? Ot should such a check be done in the caller side in RtpStreamSend::StorePacket() method?

I'm fine with overriding an older packet if needed, in `Insert()`.

**Comment by ibc** [comment]
> > Should we check current buffer size within RtpStreamSend::StorageItemBuffer::Insert() method and override an older packet if needed? Ot should such a check be done in the caller side in RtpStreamSend::StorePacket() method?
> 
> I'm fine with overriding an older packet if needed, in `Insert()`.

If we allow new packets in the buffer in a position before old packets, the whole `ClearOldPackets()` method is gonna become dramatic. To be clear: if we allow that buffer item in first position has timestamp B and there second item has timestamp A, and B > A, then everything becomes crazy and `ClearOldPackets()` is no longer gonna clear packets as expected since it `breaks` in those cases:

```c++
	void RtpStreamSend::ClearOldPackets(const RtpPacket* packet)
	{
		MS_TRACE();

		auto clockRate{ this->params.clockRate };

		const auto bufferSize = this->storageItemBuffer.GetBufferSize();

		// Go through all buffer items starting with the first and free all storage
		// items that contain packets older than this->maxRetransmissionDelayMs.
		for (size_t i{ 0 }; i < bufferSize && this->storageItemBuffer.GetBufferSize() != 0; ++i)
		{
			auto* storageItem = this->storageItemBuffer.GetFirst();
			const uint32_t diffTs{ packet->GetTimestamp() - storageItem->timestamp };

			// Processing RTP packet is older than first one.
			if (RTC::SeqManager<uint32_t>::IsSeqLowerThan(packet->GetTimestamp(), storageItem->timestamp))
			{
				break;
			}

			// First RTP packet is recent enough.
			if (static_cast<uint32_t>(diffTs * 1000 / clockRate) < this->retransmissionBufferSize)
			{
				break;
			}

			// Unfill the buffer start item.
			this->storageItemBuffer.RemoveFirst();
		}
	}
```

**Comment by jmillan** [comment]
> If we allow new packets in the buffer in a position before old packets, the whole ClearOldPackets() method is gonna become dramatic. To be clear: if we allow that buffer item in first position has timestamp B and there second item has timestamp A, and B > A, then everything becomes crazy and ClearOldPackets() is no longer gonna clear packets as expected since it breaks in those cases:

Yes, as I commented at the beginning IMHO this is a cleanup issue. We follow the RTP rules to clean up **old** packets (based on RTP timestamp). But what about players that do not correctly play RTP game like the offending ffmpeg? We are not prepared to detect those buggy endpoints. So we could detect those and properly cleanup instead of letting each `RtpStreamSend`  allocate 65535 items and override indexes on `Insert()`.

We could have an exaggerated `MAX_RTP_PACKETS_PER_FRAME = 100u` which would allow us remove older packets when we are receiving 101 packets with the same RTP timestamp. So first packet in the buffer and all the following with the same RTP timestamp would be removed when we receive the 101st RTP packet with the same timestamp.

**Comment by ibc** [comment]
I don't agree this is a "clear old packets" issue. Insert() must protect itself from storing more packets than allowed. It doesn't mean that your point is invalid but it may not fix all potential issues.

Do we agree that we must make Insert() safe first?

**Comment by jmillan** [comment]
> It doesn't mean that your point is invalid but it may not fix all potential issues.

* We play RTP game and clear **old** packets based on RTP timestamp.
* We can fix the `always same RTP timestamp` problem provoking this crash by properly detecting them too.
* We already handle mis-order.

What other scenarios may we be not handling?

> Do we agree that we must make Insert() safe first?

If this means that 1st stored RTP packet's RTP timestamp can be higher (newer) than 2nd packet's RTP timestamp then IMHO we are playing non RTP rules ourselves in our implementation and complicating things.

I would force some minimums internally, being forcing RTP timestamp to be increasing one of them. This was one of the whole reasons for replacing the `std::map` with a `std::deque` for storing the retransmission items, using just the minimum amount of memory at any moment.

**Comment by ibc** [comment]
@jmillan where do you suggest we should avoid that the buffer gets more than MaxSeq packets? We are currently not doing it neither in Insert() method nor in StorePacket() method.

Scenario? Imagine a client that sends 99 packets with proper increasing seq and same timestamp 100 and then 99 packets with timestamp 101 and so on.

**Comment by ibc** [comment]
**TODO:** Let's rethink about when `ClearOldPackets()` must be called. Currently it is called for every packet received by `RtpStreamSend` which means that, if the `Producer` is paused for 2 hours so no new RTP packets are received by their `Consumers`, we are gonna keep in the retransmission buffer 2 hours old packets. IMHO it doesn't make any sense.

I'm about to suggest using a periodic timer and clear old packets every N seconds. Yes, only problem here is that if the `Producer` sends many packets per second, the buffer may be filled soon. However remember that we plan to remove old packets when inserting new packets in the buffer to avoid the buffer to get filled.

**Comment by ibc** [comment]
Here the ongoing refactor and the new `RetransmissionBuffer::Insert()` implementation. Notes:

- `FillItem()` 

```c++
/**
 * This method tries to insert given packet into the buffer. Here we assume
 * that packet seq number is legitimate according to the content of the buffer.
 * We discard the packet if too old and also discard it if its timestamp does
 * not properly fit (by ensuring that elements in the buffer are not only
 * ordered by increasing seq but also that their timestamp are incremental).
 */
void RtpStreamSend::RetransmissionBuffer::Insert(
  RTC::RtpPacket* packet, std::shared_ptr<RTC::RtpPacket>& sharedPacket)
{
	MS_TRACE();

	auto ssrc      = packet->GetSsrc();
	auto seq       = packet->GetSequenceNumber();
	auto timestamp = packet->GetTimestamp();

	// Buffer is empty, so just insert new item.
	if (this->buffer.empty())
	{
		auto* item = new RetransmissionItem();

		this->buffer.push_back(FillItem(item, packet, sharedPacket));

		// Packet's seq number becomes startSeq.
		this->startSeq = seq;

		return;
	}

	auto* oldestItem = GetOldest();
	auto* newestItem = GetNewest();

	// Packet arrived in order (its seq is higher than seq of the newest stored
	// packet) so will become the newest one in the buffer.
	if (RTC::SeqManager<uint16_t>::IsSeqHigherThan(seq, newestItem->sequenceNumber))
	{
		// Ensure that the timestamp of the packet is equal or higher than the
		// timestamp of the newest stored packet.
		if (RTC::SeqManager<uint32_t>::IsSeqLowerThan(timestamp, newestItem->timestamp))
		{
			MS_WARN_TAG(
			  rtp,
			  "packet has higher seq but less timestamp than newest stored packet, discarding it [ssrc:%" PRIu32
			  ", seq:%" PRIu16 ", timestamp:%" PRIu32 "]",
			  ssrc,
			  seq,
			  timestamp);

			return;
		}

		// Calculate how many blank slots it would be necessary to add when
		// pushing new item to the back of the buffer.
		auto numBlankSlots = static_cast<uint16_t>(seq - newestItem->sequenceNumber - 1);

		// We may have to remove oldest items not to exceed the maximum size of
		// the buffer.
		if (this->buffer.size() + numBlankSlots + 1 > this->maxItems)
		{
			RemoveAtLeast(this->buffer.size() + numBlankSlots + 1 - this->maxItems);
		}

		// Push blank slots to the back.
		for (uint16_t i{ 0u }; i < numBlankSlots; ++i)
		{
			this->buffer.push_back(nullptr);
		}

		// Push the packet, which becomes the newest one in the buffer.
		auto* item = new RetransmissionItem();

		this->buffer.push_back(FillItem(item, packet, sharedPacket));
	}
	// Packet arrived out order and its seq is less than seq of the oldest
	// stored packet, so will become the oldest one in the buffer.
	else if (RTC::SeqManager<uint16_t>::IsSeqLowerThan(seq, oldestItem->sequenceNumber))
	{
		// Ensure that packet is not too old to be stored.
		if (IsPacketToOld(packet))
		{
			return;
		}

		// Ensure that the timestamp of the packet is equal or less than the
		// timestamp of the oldest stored packet.
		if (RTC::SeqManager<uint32_t>::IsSeqHigherThan(timestamp, oldestItem->timestamp))
		{
			MS_WARN_TAG(
			  rtp,
			  "packet has less seq but higher timestamp than oldest stored packet, discarding it [ssrc:%" PRIu32
			  ", seq:%" PRIu16 ", timestamp:%" PRIu32 "]",
			  ssrc,
			  seq,
			  timestamp);

			return;
		}

		// Calculate how many blank slots it would be necessary to add when
		// pushing new item to the fton of the buffer.
		auto numBlankSlots = static_cast<uint16_t>(oldestItem->sequenceNumber - seq - 1);

		// If adding this packet (and needed blank slots) to the front makes the
		// buffer exceed its max size, discard this packet.
		if (this->buffer.size() + numBlankSlots + 1 > this->maxItems)
		{
			MS_WARN_TAG(
			  rtp,
			  "discarding received old packet to not exceed max buffer size [ssrc:%" PRIu32
			  ", seq:%" PRIu16 ", timestamp:%" PRIu32 "]",
			  ssrc,
			  seq,
			  timestamp);

			return;
		}

		// Push blank slots to the front.
		for (uint16_t i{ 0u }; i < numBlankSlots; ++i)
		{
			this->buffer.push_front(nullptr);
		}

		// Insert the packet, which becomes the oldest one in the buffer.
		auto* item = new RetransmissionItem();

		this->buffer.push_front(FillItem(item, packet, sharedPacket));

		// Packet's seq number becomes startSeq.
		this->startSeq = seq;
	}
	// Otherwise packet must be inserted between oldest and newest stored items
	// so there is already an allocated slot for it.
	else
	{
		// Let's check if an item already exist in same position. If so, assume
		// it's duplicated.
		auto* item = Get(seq);

		if (item)
		{
			return;
		}

		// idx is the intended position of the received packet in the buffer.
		auto idx = static_cast<uint16_t>(seq - this->startSeq);

		// Validate that packet timestamp is equal or higher than the timestamp of
		// the immediate older packet (if any).
		for (auto idx2 = static_cast<int32_t>(idx - 1); idx2 >= 0; --idx2)
		{
			auto* olderItem = this->buffer.at(idx2);

			// Blank slot, continue.
			if (!olderItem)
			{
				continue;
			}

			// We are done.
			if (timestamp >= olderItem->timestamp)
			{
				break;
			}
			else
			{
				MS_WARN_TAG(
				  rtp,
				  "packet timestamp is less than timestamp of immediate older packet in the buffer, discarding it [ssrc:%" PRIu32
				  ", seq:%" PRIu16 ", timestamp:%" PRIu32 "]",
				  ssrc,
				  seq,
				  timestamp);

				return;
			}
		}

		// Validate that packet timestamp is equal or less than the timestamp of
		// the immediate newer packet (if any).
		for (auto idx2 = static_cast<uint16_t>(idx + 1); idx2 < this->buffer.size(); ++idx2)
		{
			auto* newerItem = this->buffer.at(idx2);

			// Blank slot, continue.
			if (!newerItem)
			{
				continue;
			}

			// We are done.
			if (timestamp <= newerItem->timestamp)
			{
				break;
			}
			else
			{
				MS_WARN_TAG(
				  rtp,
				  "packet timestamp is higher than timestamp of immediate newer packet in the buffer, discarding it [ssrc:%" PRIu32
				  ", seq:%" PRIu16 ", timestamp:%" PRIu32 "]",
				  ssrc,
				  seq,
				  timestamp);

				return;
			}
		}

		// Store the packet.
		item = new RetransmissionItem();

		this->buffer[idx] = FillItem(item, packet, sharedPacket);
	}

	MS_ASSERT(
	  this->buffer.size() <= this->maxItems,
	  "RetransmissionBuffer contains %" PRIu16 " items (more than %" PRIu16 " max items)",
	  this->buffer.size(),
	  this->maxItems);
}
```

Some additional private methods:

```c++
RtpStreamSend::RetransmissionItem* RtpStreamSend::RetransmissionBuffer::FillItem(
  RtpStreamSend::RetransmissionItem* item,
  RTC::RtpPacket* packet,
  std::shared_ptr<RTC::RtpPacket>& sharedPacket) const
{
	MS_TRACE();

	// Store original packet and some extra info into the item.
	item->packet         = sharedPacket;
	item->ssrc           = packet->GetSsrc();
	item->sequenceNumber = packet->GetSequenceNumber();
	item->timestamp      = packet->GetTimestamp();
	item->receivedAtMs   = DepLibUV::GetTimeMs();

	// Store original packet into the item. Only clone once and only if
	// necessary.
	if (!sharedPacket.get())
	{
		sharedPacket.reset(packet->Clone());
	}

	return item;
}
```

In addition, `RemoveAtLeast(uint16_t numItems)` is not yet implemented but it will basically remove, at least, `numItems` items from the buffer front plus, additionally, blank slots remaining at the front.

File: worker/include/RTC/RtpStream.hpp
Comment (ibc) [inline worker/include/RTC/RtpStream.hpp:1]: Cosmetic changes in `RtpStream.hpp`. Let's avoid those long multi-line comments at the right side of class members.

File: worker/include/RTC/RtpStreamRecv.hpp
Comment (ibc) [inline worker/include/RTC/RtpStreamRecv.hpp:1]: Same. Cosmetic changes in `RtpStreamRecv.hpp`. Let's avoid those long multi-line comments at the right side of class members.

File: worker/include/common.hpp
Comment (ibc) [inline worker/include/common.hpp:1]: Cosmetic changes in comments. Ignore.

File: worker/src/RTC/RetransmissionBuffer.cpp
Comment (ibc) [inline worker/src/RTC/RetransmissionBuffer.cpp:1]: A separate class.

File: worker/src/RTC/RtpStreamRecv.cpp
Comment (ibc) [inline worker/src/RTC/RtpStreamRecv.cpp:1]: Cosmetic. Ignore.

File: worker/src/handles/Timer.cpp
Comment (ibc) [inline worker/src/handles/Timer.cpp:1]: Cosmetic. Ignore.

**Comment by ibc** [comment]
@jmillan @nazar-pc this is ready for review.

File: worker/src/RTC/RtpStreamSend.cpp [hunk 1]
```diff
@@ -11,163 +11,19 @@ namespace RTC
 {
 	/* Static. */
 
+	// Limit retransmission buffer max size to 2500 items.
+	static constexpr size_t RetransmissionBufferMaxItems{ 2500u };
```
Comment (ibc) [inline worker/src/RTC/RtpStreamSend.cpp:15]: This is the retransmission buffer max size (passed to `RetransmissionBuffer` class as argument in the constructor. In my tests this is good enough.

**Comment by ibc** [comment]
`TestRetransmissionBuffer.cpp` added!

File: worker/src/RTC/RetransmissionBuffer.cpp [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+		}
+
+		// If we emptied the full buffer, reset startSeq.
+		if (this->buffer.empty())
+		{
+			this->startSeq = 0u;
+		}
+	}
+
+	void RetransmissionBuffer::RemoveFromFrontAtLeast(uint16_t numItems)
```
Comment (jmillan) [inline worker/src/RTC/RetransmissionBuffer.cpp]: for the same reason `RemoveOldest` is not called `RemoveOldestAtLeast` I would rename this method to `RemoveFromFront`

**Review by jmillan — APPROVED** [review]
Nice refactor :+1:

File: worker/src/RTC/RetransmissionBuffer.cpp (hunk 2 above)
Comment (ibc) [reply worker/src/RTC/RetransmissionBuffer.cpp]: What about RemoveOldest(numItems) instead? This is, a different signature

Comment (ibc) [reply worker/src/RTC/RetransmissionBuffer.cpp]: I've done it and pushed it.

**Comment by vpalmisano** [comment]
After upgrading to this version, we received a SIGABRT, with this stack trace:
```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x0000ffffabab7aa0 in __GI_abort () at abort.c:79
#2  0x0000ffffabdad238 in __gnu_cxx::__verbose_terminate_handler() () from /usr/lib/aarch64-linux-gnu/libstdc++.so.6
#3  0x0000ffffabdaad4c in ?? () from /usr/lib/aarch64-linux-gnu/libstdc++.so.6
#4  0x0000ffffabdaadb0 in std::terminate() () from /usr/lib/aarch64-linux-gnu/libstdc++.so.6
#5  0x0000ffffabdab0a4 in __cxa_throw () from /usr/lib/aarch64-linux-gnu/libstdc++.so.6
#6  0x0000ffffabdd3950 in std::__throw_out_of_range_fmt(char const*, ...) () from /usr/lib/aarch64-linux-gnu/libstdc++.so.6
#7  0x0000aaaabc66aa3c in RTC::RetransmissionBuffer::Insert(RTC::RtpPacket*, std::shared_ptr<RTC::RtpPacket>&) ()
#8  0x0000aaaabc68aadc in RTC::RtpStreamSend::ReceivePacket(RTC::RtpPacket*, std::shared_ptr<RTC::RtpPacket>&) ()
#9  0x0000aaaabc6afca4 in RTC::SvcConsumer::SendRtpPacket(RTC::RtpPacket*, std::shared_ptr<RTC::RtpPacket>&) ()
#10 0x0000aaaabc676384 in RTC::Router::OnTransportProducerRtpPacketReceived(RTC::Transport*, RTC::Producer*, RTC::RtpPacket*) ()
#11 0x0000aaaabc663e7c in RTC::Producer::ReceiveRtpPacket(RTC::RtpPacket*) ()
#12 0x0000aaaabc6b4b18 in RTC::Transport::ReceiveRtpPacket(RTC::RtpPacket*) ()
#13 0x0000aaaabc64b0d8 in RTC::PipeTransport::OnUdpSocketPacketReceived(RTC::UdpSocket*, unsigned char const*, unsigned long, sockaddr const*) ()
#14 0x0000aaaabc96cc34 in uv.udp_recvmmsg ()
#15 0x0000aaaabc96d8c4 in uv.udp_io ()
#16 0x0000aaaabc970cc4 in uv.io_poll ()
#17 0x0000aaaabc9646a4 in uv_run ()
#18 0x0000aaaabc5c6074 in DepLibUV::RunLoop() ()
#19 0x0000aaaabc5cf3d0 in Worker::Worker(Channel::ChannelSocket*, PayloadChannel::PayloadChannelSocket*) ()
#20 0x0000aaaabc5c4c14 in mediasoup_worker_run ()
#21 0x0000aaaabc5c3834 in main ()
```

**Comment by ibc** [comment]
@vpalmisano please create a separate issue for this.

**Pull request merged**